File: internal/graphql/lint/alphabetize.go
// Copyright 2026 Outreach Corporation. All Rights Reserved.

// Description: Tier 3 alphabetize rule.

// alphabetize.go implements 1 (of 10) Tier 3 custom rules: alphabetize,
// which requires alphabetical order among the fields, enum values,
// and/or arguments its "fields", "values", and "arguments" options
// select, matching @graphql-eslint's rule of the same name.
//
// devbase graphql lint only ever parses schema files (see lint.go's
// Files), so this port narrows @graphql-eslint's own option surface to
// its "schema" configOptions -- "fields", "values", and "arguments" --
// dropping "selections" and "variables" (operation-only) and
// "definitions" (a v4-only option that was already a no-op in every
// repo's config carried into this migration; see RFC 0006's Tier 3
// table).
//
// A type extension's fields or enum values are ordered together with
// its base type's own, via allDefinitions (tier3.go):
// validator.ValidateSchemaDocument already merges an extension's Fields
// and EnumValues into the base Definition's, so this rule sees one
// alphabetical run across a type's base definition and every extension
// of it, wherever each is written. @graphql-eslint itself cannot do
// this: it lints one file's own AST at a time, so it only orders a base
// definition's own fields against each other, and separately orders
// each extension's own fields against each other.

package lint

import (
	"strings"
	"unicode"

	"github.com/getoutreach/devbase/v2/internal/graphql/config"
	"github.com/getoutreach/gobox/pkg/set"
	"github.com/vektah/gqlparser/v2/ast"
	"github.com/vektah/gqlparser/v2/gqlerror"
)

// alphabetizeFieldKinds maps alphabetize's "fields" option entries to
// the ast.DefinitionKind each names, restricted to the 3 kinds
// @graphql-eslint's own fieldsEnum allows.
//
//nolint:gochecknoglobals // Why: a fixed lookup table, never mutated; Go has no map consts.
var alphabetizeFieldKinds = map[string]ast.DefinitionKind{
	"ObjectTypeDefinition":      ast.Object,
	"InterfaceTypeDefinition":   ast.Interface,
	"InputObjectTypeDefinition": ast.InputObject,
}

// alphabetizeOptions is alphabetize's scripts/devbase.yaml options,
// decoded from Rule.Options.
type alphabetizeOptions struct {
	// fieldKinds is which of Object, Interface, and InputObject the
	// "fields" option selects.
	fieldKinds set.Set[ast.DefinitionKind]

	// values is whether the "values" option selects EnumTypeDefinition
	// -- the only value @graphql-eslint's own valuesEnum allows.
	values bool

	// fieldArguments and directiveArguments are whether the
	// "arguments" option selects FieldDefinition and DirectiveDefinition,
	// respectively.
	fieldArguments     bool
	directiveArguments bool
}

// parseAlphabetizeOptions decodes opts (Rule.Options) into
// alphabetizeOptions. Any option absent from opts, or opts itself nil,
// leaves the corresponding check disabled, matching @graphql-eslint's
// own behavior of a kind being excluded until its option names it.
func parseAlphabetizeOptions(opts map[string]any) alphabetizeOptions {
	o := alphabetizeOptions{fieldKinds: set.Of[ast.DefinitionKind]()}
	for _, key := range stringSliceOption(opts, "fields") {
		if kind, ok := alphabetizeFieldKinds[key]; ok {
			o.fieldKinds.Insert(kind)
		}
	}
	for _, key := range stringSliceOption(opts, "values") {
		if key == typeDefinitionKindKeys[ast.Enum] {
			o.values = true
		}
	}
	for _, key := range stringSliceOption(opts, "arguments") {
		switch key {
		case fieldDefinitionSelector:
			o.fieldArguments = true
		case directiveDefinitionSelector:
			o.directiveArguments = true
		}
	}
	return o
}

// alphaName is one name alphabetize compares against its predecessor:
// a field, enum value, or argument, at the position its violation (if
// any) should be reported at.
type alphaName struct {
	name string
	pos  *ast.Position
}

// inScopeNames filters names to those written in one of inScope's
// sources, dropping a nil Position (an injected introspection
// meta-field; see descriptions.go) the same way.
func inScopeNames(names []alphaName, inScope set.Set[*ast.Source]) []alphaName {
	out := make([]alphaName, 0, len(names))
	for _, n := range names {
		if n.pos != nil && inScope.Contains(n.pos.Src) {
			out = append(out, n)
		}
	}
	return out
}

// localeCompare mimics JavaScript's default String.prototype.localeCompare
// closely enough for alphabetize's own comparisons, which @graphql-eslint
// itself uses to order names: case-insensitive at the primary level, so
// names group by base letter regardless of case, with a
// lowercase-before-uppercase tiebreak when two names are
// case-insensitively equal -- the opposite of plain byte order, where an
// uppercase letter sorts before its lowercase counterpart.
func localeCompare(a, b string) int {
	if la, lb := strings.ToLower(a), strings.ToLower(b); la != lb {
		if la < lb {
			return -1
		}
		return 1
	}

	ra, rb := []rune(a), []rune(b)
	for i := 0; i < len(ra) && i < len(rb); i++ {
		if ra[i] == rb[i] {
			continue
		}
		if aLower, bLower := unicode.IsLower(ra[i]), unicode.IsLower(rb[i]); aLower != bLower {
			if aLower {
				return -1
			}
			return 1
		}
		if ra[i] < rb[i] {
			return -1
		}
		return 1
	}

	switch {
	case len(ra) < len(rb):
		return -1
	case len(ra) > len(rb):
		return 1
	default:
		return 0
	}
}

// checkAlphaOrder reports a RuleAlphabetize violation for every name in
// names that sorts after its successor per localeCompare.
func checkAlphaOrder(names []alphaName) []Violation {
	var violations []Violation
	for i := 1; i < len(names); i++ {
		prev, curr := names[i-1], names[i]
		if localeCompare(prev.name, curr.name) == 1 {
			violations = append(violations, Violation{
				err:  gqlerror.ErrorPosf(curr.pos, "`%s` should be before `%s`.", curr.name, prev.name),
				Rule: config.RuleAlphabetize,
			})
		}
	}
	return violations
}

// fieldNames returns fields as alphaNames, in order.
func fieldNames(fields ast.FieldList) []alphaName {
	names := make([]alphaName, len(fields))
	for i, f := range fields {
		names[i] = alphaName{name: f.Name, pos: f.Position}
	}
	return names
}

// enumValueNames returns values as alphaNames, in order.
func enumValueNames(values ast.EnumValueList) []alphaName {
	names := make([]alphaName, len(values))
	for i, v := range values {
		names[i] = alphaName{name: v.Name, pos: v.Position}
	}
	return names
}

// argumentNames returns args as alphaNames, in order.
func argumentNames(args ast.ArgumentDefinitionList) []alphaName {
	names := make([]alphaName, len(args))
	for i, a := range args {
		names[i] = alphaName{name: a.Name, pos: a.Position}
	}
	return names
}

// alphabetizeViolations reports every RuleAlphabetize violation in defs
// and directives that opts selects. inScope restricts every check to
// names actually written in one of the repository's own files.
func alphabetizeViolations(defs []scopedDefinition, directives ast.DirectiveDefinitionList,
	inScope set.Set[*ast.Source], opts alphabetizeOptions,
) []Violation {
	var violations []Violation

	for _, sd := range defs {
		def := sd.def
		if opts.fieldKinds.Contains(def.Kind) {
			violations = append(violations, checkAlphaOrder(inScopeNames(fieldNames(def.Fields), inScope))...)
		}
		if opts.values && def.Kind == ast.Enum {
			violations = append(violations, checkAlphaOrder(inScopeNames(enumValueNames(def.EnumValues), inScope))...)
		}
		if opts.fieldArguments && (def.Kind == ast.Object || def.Kind == ast.Interface) {
			for _, f := range def.Fields {
				violations = append(violations, checkAlphaOrder(inScopeNames(argumentNames(f.Arguments), inScope))...)
			}
		}
	}

	if opts.directiveArguments {
		for _, dd := range directives {
			violations = append(violations, checkAlphaOrder(inScopeNames(argumentNames(dd.Arguments), inScope))...)
		}
	}

	return violations
}

// tier3Alphabetize runs alphabetize against defs. It does not run
// unless cfg enables it (config.Lint.Enabled).
func tier3Alphabetize(defs []scopedDefinition, parsed *parsedSchema, inScope set.Set[*ast.Source],
	cfg *config.Lint,
) []Violation {
	if !cfg.Enabled(config.RuleAlphabetize) {
		return nil
	}
	opts := parseAlphabetizeOptions(cfg.Options(config.RuleAlphabetize))
	return alphabetizeViolations(defs, parsed.doc.Directives, inScope, opts)
}

File: graphql-lint/crates/gql_lint_rules/src/alphabetize.rs
//! Ports `internal/graphql/lint/alphabetize.go`'s `alphabetize` rule:
//! requires alphabetical order among the fields, enum values, and/or
//! arguments its options select.
//!
//! Matches the Go port's own narrowing of `@graphql-eslint`'s option
//! surface to schema-only kinds (`ObjectTypeDefinition`,
//! `InterfaceTypeDefinition`, `InputObjectTypeDefinition` for `fields`;
//! `EnumTypeDefinition` for `values`; field and directive arguments for
//! `arguments`) — this tool only ever lints schema files, never operations,
//! so the operation-only and v4-only options the Go port already dropped
//! stay dropped here too.
//!
//! One difference from the Go port worth calling out, not hiding: the Go
//! version orders a type's base definition and every extension of it
//! together in one alphabetical run, because `gqlparser` merges them
//! before validation. `apollo-compiler` does the same folding (each
//! `ExtendedType` variant's doc comment says so explicitly: "with all
//! information from type extensions folded in"), so this port gets the
//! same merged-run behavior for free, not as separately-ported logic.

use apollo_compiler::Schema;
use apollo_compiler::schema::{ExtendedType, Name};
use gql_lint_core::{Violation, rules};

/// Which of `fields`'s three schema-only kinds are selected. Split out of
/// [`AlphabetizeOptions`] into its own small group (clippy's
/// `struct_excessive_bools` pedantic lint rejects a flat struct with more
/// than 3 bare `bool` fields; grouping by the option key they came from is
/// also just a clearer shape than 6 flat, same-typed fields).
#[derive(Debug, Clone, Copy, Default)]
pub struct FieldKinds {
    pub object: bool,
    pub interface: bool,
    pub input_object: bool,
}

/// Which of `arguments`'s two kinds are selected.
#[derive(Debug, Clone, Copy, Default)]
pub struct ArgumentKinds {
    pub field: bool,
    pub directive: bool,
}

/// Which of `alphabetize`'s options are enabled, decoded from
/// `scripts/devbase.yaml`'s `graphql.lint.rules.alphabetize` options —
/// mirrors the Go port's `alphabetizeOptions`.
#[derive(Debug, Clone, Copy, Default)]
pub struct AlphabetizeOptions {
    pub fields: FieldKinds,
    pub enum_values: bool,
    pub arguments: ArgumentKinds,
}

impl AlphabetizeOptions {
    /// Decodes options from the raw `{fields: [...], values: [...],
    /// arguments: [...]}` mapping `scripts/devbase.yaml` provides, the same
    /// shape the Go port's `parseAlphabetizeOptions` reads.
    #[must_use]
    pub fn from_yaml(options: Option<&serde_yaml::Mapping>) -> Self {
        let Some(options) = options else {
            return Self::default();
        };

        let mut opts = Self::default();
        for key in string_list(options, "fields") {
            match key {
                "ObjectTypeDefinition" => opts.fields.object = true,
                "InterfaceTypeDefinition" => opts.fields.interface = true,
                "InputObjectTypeDefinition" => opts.fields.input_object = true,
                _ => {}
            }
        }
        for key in string_list(options, "values") {
            if key == "EnumTypeDefinition" {
                opts.enum_values = true;
            }
        }
        for key in string_list(options, "arguments") {
            match key {
                "FieldDefinition" => opts.arguments.field = true,
                "DirectiveDefinition" => opts.arguments.directive = true,
                _ => {}
            }
        }
        opts
    }
}

fn string_list<'a>(options: &'a serde_yaml::Mapping, key: &str) -> Vec<&'a str> {
    options
        .get(key)
        .and_then(serde_yaml::Value::as_sequence)
        .map(|seq| seq.iter().filter_map(serde_yaml::Value::as_str).collect())
        .unwrap_or_default()
}

/// Runs the `alphabetize` rule over `schema`, per `opts`.
///
/// `fields` and `arguments` are independent options, matching the Go
/// port's own structure (two separate `if` conditions, not one nested
/// inside the other) — a repo can enable `arguments: [FieldDefinition]`
/// without also enabling `fields: [ObjectTypeDefinition]`, and each
/// object/interface type's field order and its fields' argument order
/// must be checked independently, not gated on each other.
///
/// **Caught by running this against a real fixture, not by inspection:**
/// `Schema::builder()` pre-populates built-in scalars and the reserved
/// `__`-prefixed introspection types (`__Schema`, `__Type`, `__Field`, and
/// so on). An early version of this function walked `schema.types`
/// without excluding them and reported dozens of spurious violations
/// against GraphQL's own introspection schema, alongside the one real
/// violation in the fixture. The Go port avoids this by restricting every
/// Tier 3 walk to `inScope`, definitions whose source file identity
/// matches one of the repository's own `*.graphql` files
/// (`internal/graphql/lint/tier3.go`). `apollo-compiler` exposes a
/// simpler, direct equivalent: [`ExtendedType::is_built_in`].
#[must_use]
pub fn alphabetize(schema: &Schema, opts: &AlphabetizeOptions) -> Vec<Violation> {
    let mut violations = Vec::new();

    for ty in schema.types.values().filter(|ty| !ty.is_built_in()) {
        match ty {
            ExtendedType::Object(t) => {
                if opts.fields.object {
                    check_order(schema, &mut violations, t.fields.keys());
                }
                if opts.arguments.field {
                    for field in t.fields.values() {
                        check_order(
                            schema,
                            &mut violations,
                            field.arguments.iter().map(|a| &a.name),
                        );
                    }
                }
            }
            ExtendedType::Interface(t) => {
                if opts.fields.interface {
                    check_order(schema, &mut violations, t.fields.keys());
                }
                if opts.arguments.field {
                    for field in t.fields.values() {
                        check_order(
                            schema,
                            &mut violations,
                            field.arguments.iter().map(|a| &a.name),
                        );
                    }
                }
            }
            ExtendedType::InputObject(t) if opts.fields.input_object => {
                check_order(schema, &mut violations, t.fields.keys());
            }
            ExtendedType::Enum(t) if opts.enum_values => {
                check_order(schema, &mut violations, t.values.keys());
            }
            _ => {}
        }
    }

    if opts.arguments.directive {
        for directive in schema
            .directive_definitions
            .values()
            .filter(|d| !d.is_built_in())
        {
            check_order(
                schema,
                &mut violations,
                directive.arguments.iter().map(|a| &a.name),
            );
        }
    }

    violations
}

/// Reports a violation for every name in `names` that sorts after its
/// successor per [`locale_compare`] — the Rust port of the Go rule's
/// `checkAlphaOrder`.
fn check_order<'a>(
    schema: &Schema,
    violations: &mut Vec<Violation>,
    names: impl Iterator<Item = &'a Name>,
) {
    let mut prev: Option<&Name> = None;
    for curr in names {
        if let Some(prev) = prev
            && locale_compare(prev.as_str(), curr.as_str()) == std::cmp::Ordering::Greater
        {
            let (file, line, column) =
                gql_lint_core::resolve_location(&schema.sources, curr.location());
            violations.push(Violation {
                file,
                line,
                column,
                message: format!("`{curr}` should be before `{prev}`."),
                rule: rules::ALPHABETIZE,
            });
        }
        prev = Some(curr);
    }
}

/// Mimics JavaScript's default `String.prototype.localeCompare` closely
/// enough for `alphabetize`'s own comparisons, matching
/// `@graphql-eslint`'s ordering exactly: case-insensitive at the primary
/// level (so names group by base letter regardless of case), with a
/// lowercase-before-uppercase tiebreak when two names are
/// case-insensitively equal — the opposite of plain byte order, where an
/// uppercase letter sorts before its lowercase counterpart. A direct port
/// of the Go rule's own `localeCompare`, kept because the ordering it
/// produces is part of this rule's observable behavior, not an
/// implementation detail free to pick idiomatically.
fn locale_compare(a: &str, b: &str) -> std::cmp::Ordering {
    let (lower_a, lower_b) = (a.to_lowercase(), b.to_lowercase());
    if lower_a != lower_b {
        return lower_a.cmp(&lower_b);
    }

    for (ca, cb) in a.chars().zip(b.chars()) {
        if ca == cb {
            continue;
        }
        let (a_lower, b_lower) = (ca.is_lowercase(), cb.is_lowercase());
        if a_lower != b_lower {
            return if a_lower {
                std::cmp::Ordering::Less
            } else {
                std::cmp::Ordering::Greater
            };
        }
        return ca.cmp(&cb);
    }

    a.chars().count().cmp(&b.chars().count())
}

#[cfg(test)]
mod tests {
    use super::locale_compare;
    use std::cmp::Ordering;

    #[test]
    fn case_insensitive_primary_order() {
        assert_eq!(locale_compare("apple", "Banana"), Ordering::Less);
        assert_eq!(locale_compare("Banana", "apple"), Ordering::Greater);
    }

    #[test]
    fn lowercase_before_uppercase_tiebreak() {
        // Case-insensitively equal ("id" == "id"), differ only by case:
        // lowercase must sort first, the reverse of plain byte order.
        assert_eq!(locale_compare("id", "Id"), Ordering::Less);
        assert_eq!(locale_compare("Id", "id"), Ordering::Greater);
    }

    #[test]
    fn equal_names() {
        assert_eq!(locale_compare("name", "name"), Ordering::Equal);
    }
}
